### PR TITLE
Target includes: #include "osqp.h" -> #include "osqp/osqp.h"

### PR DESCRIPTION
**Pull request by mhubii**
This PR aims to change the `osqp` target's include directories from `include/osqp` to `include` (refers to #400), see for example

https://github.com/osqp/osqp/blob/33454b3e236f1f44193bfbbb6b8c8e71f8f04e9a/CMakeLists.txt#L285

This way, `osqp` is included via `#include "osqp/osqp.h"` as opposed to `#include "osqp.h"`. 

This shall help remove possible ambiguity. The tests that I ran locally passed.

I am aware of the impact to plenty downstream libraries, please let me know what you think, nonetheless. 

Thanks! Kind regards

### CODE REVIEW

**Comment by mhubii** [comment]
just saw in another PR that PRs are to be targeted at branch `develop-1.0`. I'll try to replicate this behavior there. In `develop-1.0`, there is indeed a very uncommon way to separate public and private headers.

**Comment by imciner2** [comment]
As mentioned in the other PR, we are currently focused on development on the `develop-1.0` branch instead of master, and this change would be too disruptive for inclusion in a standard release before 1.0 (because it would break the includes for any project using OSQP). I also don't think this will be needed in the 1.0 branch, because we have already cleaned up the includes to separate the public and private includes, and all but one include starts with `osqp_` in its name to namespace it into OSQP.

**Comment by mhubii** [comment]
yes, I have seen this, sorting the headers in this manner is really uncommon though.

**Comment by imciner2** [comment]
There is no standard way of organizing a C code base - every project I have ever examined organizes theirs differently, and this is the way we have chosen to structure this project. A consumer of OSQP should only use the headers contained in the `public` folder and should not use any headers in the `private` folder. The contents of the `public` folder will be installed into `$<INSTALL_PREFIX>/${CMAKE_INSTALL_INCLUDEDIR}/osqp`, and the private headers won't. So the main OSQP header will be located at `$<INSTALL_PREFIX>/${CMAKE_INSTALL_INCLUDEDIR}/osqp/osqp.h` when installed.

Any project using OSQP through CMake will be able to get the header locations from CMake in the standard way. If they include OSQP as a subdirectory directly, then just adding a target dependency against either `osqpstatic` or `osqp` will give the public headers located in the source tree for the consumer to use. I am also in the process of updating the CMake to export config snippets, which will then export targets `osqp::osqp` and `osqp::osqpstatic` which can be used to define dependencies in the normal CMake way and get the include paths for the installed headers. Then all the consumer has to do is add `#include <osqp.h>` to their code, and OSQP will be used.

**Comment by mhubii** [comment]
thanks for explaining, but I understand what is happening. This really poses an issue, believe it or not. Feel free to close. Thanks nonetheless.

**Comment by imciner2** [comment]
The 1.0 branch is a breaking release already, with the actual solver API changing to remove functions and consolidate things (and rename several types), so the user disruption of a change in the 1.0 branch is a minimal factor. My previous comment was about making this change in a pre-1.0 release - where this change would be a breaking change for downstream users. But in terms of the headers, the only header previously installed by OSQP that is remaining public and will be installed in the 1.0 release is `osqp.h`, every other header in the pre-1.0 versions has become private and will not be included. The main include header was, and still remains, `osqp.h`, with it including all the other headers needed for user code.

> This really poses an issue, believe it or not.

Then please explain exactly what the issue is and provide the details needed to reproduce it. I have been responding to the issues that I pick out of all the comments, but so far I have only found two, which are fixed by the namespacing and the proper CMake target generation and described below.

The main issue I recall ever seeing mentioned in any of these threads was about a conflict between `error.h` in the OSQP library and the system `error.h`. Since we set the include paths in the library build to prioritize the OSQP headers first, I assumed this error was happening inside user code including the OSQP headers (but no response to the questions in that initial thread was provided to support or refute this idea). This issue would then be fixed by the namespacing of all our installed headers inside `osqp_` and the separation of the public/private headers so that there was no way for our installed headers to clash with other headers on the system.

The next issue I could find was the statement that to use OSQP downstream consumers had to use "old cmake standards" - which I said will not be the case for 1.0 because we are/will be doing 2 things:
* Creating targets that consumers of OSQP in-tree can simply add a target dependency to and have the proper linking and include directories
* Creating an `osqp-config.cmake` file to export the targets when the library is installed, allowing consumers to just do a `find_package` and then add a target dependency on the OSQP targets to get the right include and linking information.

Using the CMake targets is the way modern CMake is supposed to work - so there will be no need for "old cmake standards" to be used with OSQP.

In short, please provide more details on exactly what issues these changes are meant to solve and how to reproduce them so we can look at them in more detail.

**Comment by traversaro** [comment]
> I am referring to an issue that occurs in [osqp-eigen](https://github.com/robotology/osqp-eigen) when linking against `osqp` through the [osqp_vendor](https://github.com/tier4/osqp_vendor) package in ROS2. This could be resolved by exporting `include` instead of `include/osqp`. Please refer to [tier4/osqp_vendor#10](https://github.com/tier4/osqp_vendor/pull/10), [tier4/osqp_vendor#10 (comment)](https://github.com/tier4/osqp_vendor/pull/10#issuecomment-1063139034). I think the issue is rather specific, hence it might not be worth your consideration. I just thought structuring `osqp` in this manner might be beneficial overall.

Hi @mhubii, just commenting also here (beside https://github.com/tier4/osqp_vendor/pull/10/files#r823983449) as you mention `osqp-eigen`: I may be wrong, but the issue seems related to the fact that you are not using so-called "modern" CMake targets instead of "legacy" variables such as `<pkg>_INCLUDE_DIRS` and `<pkg>_LIBRARIES`.

**Comment by traversaro** [comment]
> I am referring to an issue that occurs in [osqp-eigen](https://github.com/robotology/osqp-eigen) when linking against `osqp` through the [osqp_vendor](https://github.com/tier4/osqp_vendor) package in ROS2. This could be resolved by exporting `include` instead of `include/osqp`. Please refer to [tier4/osqp_vendor#10](https://github.com/tier4/osqp_vendor/pull/10), [tier4/osqp_vendor#10 (comment)](https://github.com/tier4/osqp_vendor/pull/10#issuecomment-1063139034). I think the issue is rather specific, hence it might not be worth your consideration. I just thought structuring `osqp` in this manner might be beneficial overall.

Hi @mhubii, just commenting also here (beside https://github.com/tier4/osqp_vendor/pull/10/files#r823983449) as you mention `osqp-eigen`: I may be wrong, but the issue seems related to the fact that you are not using so-called "modern" CMake targets instead of "legacy" variables such as `<pkg>_INCLUDE_DIRS` and `<pkg>_LIBRARIES`.

**Comment by mhubii** [comment]
Hi @traversaro, Ill investigate, thanks for the hint

**Comment by traversaro** [comment]
> Hi @traversaro, Ill investigate, thanks for the hint, however, I am not using legacy variables.

Not directly, but you are using `ament_export_include_directories` that internally does. See https://github.com/ament/ament_cmake/issues/365 for a related discussion on the `ament_cmake` side.

**Comment by mhubii** [comment]
Ill check later, thx

**Comment by mhubii** [comment]
I am closing this PR and the accompanying issue #400, because I think it is up to you @imciner2 to decide how to export the include directories, and the ROS2 `osqp_vendor` package, being the consumer of `osqp`, has to adapt, not the other way around. I still believe there might be some benefit of exporting `include` over pre-fixing with `osqp_`, but again, this is up to you. Thank you for your support and the valuable insights. Kind regards.
